src/core/vmm.d

```diff
@@ -0,0 +1,489 @@
+    Important for data-heavy applications. The permissions are set to read/write.
+    See $$(LREF reserve) and $(LREF commit).)
+    $(LI Locking a range of memory pages to persist in physical memory
+        (avoid being paging out by OS). See $(LREF lock) and $(LREF unlock).)
+
+    $(P API provides a way to perform less common operations but may have some overhead
+    compared to writing platform-specific code.)
+
+*/
+
+import core.thread;
+
+nothrow @nogc:
+
+/// Flags that specify desired kind of access to the memory.
+enum MemoryAccess : uint
+{
+    none =      0b0000_0000,
+    read =      0b0000_0001,
+    write =     0b0000_0010,
+    execute =   0b0000_0100,
+    readWrite = read | write,
+    readWriteExecute = read | write | execute,
+};
+
+/**
+    Flags that specify additional memory options.
+    BUGS: None of these are implemented yet.
+*/
+enum MemoryOptions : uint
+{
+    none =      0b0000_0000,
+    hugePages = 0b0001_0000
+};
+
+/**
+    System's virtual memory manager capabilities flags,
+    indicate whether an operation is supported natively.
+    If some operation is not supported natively, the "next best thing" is done where applicable
+    such as e.g. madvising memory as "won't use" instead of forcibly decommiting pages on FreeBSD.
+*/
+enum MemoryCapability
+{
+    commit = 0x1,
+    decommit = 0x2,
+    reset = 0x4,
+    hugePages = 0x8,
+    sharedMapping = 0x10
+};
+
+version(CoreDdoc)
+    enum MemoryCapability memoryCapability =
+        MemoryCapability.commit | MemoryCapability.decommit | MemoryCapability.reset
+        | MemoryCapability.hugePages | MemoryCapability.sharedMapping;
+else version(Windows)
+    enum MemoryCapability memoryCapability =
+        MemoryCapability.commit | MemoryCapability.decommit | MemoryCapability.reset;
+else version(linux)
+    enum MemoryCapability memoryCapability =
+        MemoryCapability.commit | MemoryCapability.decommit;
+else version(FreeBSD)
+    enum MemoryCapability memoryCapability =
+        MemoryCapability.commit | MemoryCapability.reset;
+else version(OSX)
+    enum MemoryCapability memoryCapability =
+        MemoryCapability.commit | MemoryCapability.reset;
+else
+    static assert(false, "Don't know the memory capability of this OS");
+
+/// The page size as reported by the OS at the application start.
+public alias pageSize = PAGESIZE;
+
+//TODO: provide the same for huge page size
+
+/**
+    Allocate a block of $(D size) bytes of virtual memory rounded up to
+    nearest multiple of page size. The memory is allocated
+    with specified permissions and options.
+
+    $(D base) specifies the desired base address of the block, $(D null)
+    means any base address.
+
+    To release a block of memory allocated by this function use $(LREF free).
+
+    Note: This operation commits virtual memory.
+    To have fine grained control over reserve and commit phases
+    use $(LREF reserve) and $(LREF commit).
+*/
+void[] allocate(void* base, size_t size,
```

Comment:
Meta comment: no `Returns:` documentation blocks anywhere. No indication of what happens when the OS function call fails. No ability to determine why it failed (when it does fail).